### DISCUSSION TITLE
TypeError: require.resolveWeak is not a function

### DISCUSSION THREAD

**Discussion by Emiliano-Bucci** [vercel/next.js#18855] [Help]
Hi! I've recently (with the latest version of next js) start to face issues during the test (i use `jest` and `ts-jest`) of dynamic components. I start to get `TypeError: require.resolveWeak is not a function`. I've also tried to downgrade to version 8 of next but i'm still facing the issue, so i guess i'm doing something wrong (maybe inside the jest.config.js?). Here's a repo example -> https://github.com/Emiliano-Bucci/nextjs-starter

Thanks :)

**Comment by hellno** [comment]
same here :) any suggestions how to fix this for 10.0.3?

**Comment by sandcastle** [comment]
We are also seeing this issue with `v10` using `babel-jest`.

**Comment by Emiliano-Bucci** [comment]
@lfades 
None? :/

**Comment by lfades** [comment]
Can you test our Jest examples and see if they work? 🙏  If they don't then please open an issue 👍

**Reply by wpatter6** [reply]
Is there a link to the example you're talking about? I don't see any dynamic imports in the [with-jest](https://github.com/vercel/next.js/tree/canary/examples/with-jest) example.

**Comment by idoshamun** [comment]
In the meantime, here's a workaround I found:

Add this mock to your jest setup:
```
jest.mock('next/dynamic', () => (func: () => Promise<any>) => {
  let component: any = null;
  func().then((module: any) => {
    component = module.default;
  });
  const DynamicComponent = (...args) => component(...args);
  DynamicComponent.displayName = 'LoadableComponent';
  DynamicComponent.preload = jest.fn();
  return DynamicComponent;
});
```
and this to your babelrc
```
"env": {
    "test": {
      "plugins": [
        "dynamic-import-node"
      ]
    }
  }
```

**Reply by steelx** [reply]
Thanks this solves issue.

**Reply by Yimiprod** [reply]
Thanks. I took that to make my own (since we still have class component in our codebase, calling has a function threw some errors, fixed it by using `<component {...args} />`

**Comment by wmcb91** [comment]
If anyone else landed here and is working with Next at or above v12.0.8, it appears @idoshamun's solution may temporarily not work. Tracking [this issue](https://github.com/vercel/next.js/issues/33476#issuecomment-1025877866), it sounds like a solution is being worked on.

**Comment by mskashef** [comment]
Add this mock:
```js
jest.mock('next/dynamic', () => ({
  __esModule: true,
  default: (...props) => {
    const dynamicModule = jest.requireActual('next/dynamic')
    const dynamicActualComp = dynamicModule.default
    const RequiredComponent = dynamicActualComp(props[0])
    RequiredComponent.preload
      ? RequiredComponent.preload()
      : RequiredComponent.render.preload()
    return RequiredComponent
  }
}))
```
